File: common/src/main/java/com/ssblur/scriptor/helpers/SpellbookHelper.java
package com.ssblur.scriptor.helpers;

import com.ssblur.scriptor.ScriptorGameRules;
import com.ssblur.scriptor.advancement.ScriptorAdvancements;
import com.ssblur.scriptor.data.DictionarySavedData;
import com.ssblur.scriptor.helpers.targetable.SpellbookTargetable;
import com.ssblur.scriptor.word.Spell;
import net.minecraft.nbt.Tag;
import net.minecraft.network.chat.Component;
import net.minecraft.server.level.ServerLevel;
import net.minecraft.server.level.ServerPlayer;
import net.minecraft.sounds.SoundEvents;
import net.minecraft.sounds.SoundSource;
import net.minecraft.world.entity.player.Player;
import net.minecraft.world.item.Item;
import net.minecraft.world.item.ItemStack;

import java.util.ArrayList;
import java.util.List;

public class SpellbookHelper {
  public static List<Item> SPELLBOOKS = new ArrayList<>();

  public static boolean castFromItem(ItemStack itemStack, Player player) {
    var compound = itemStack.getTag();
    var level = player.level();
    if(compound == null || !(level instanceof ServerLevel server))
      return false;

    level.playSound(null, player.blockPosition(), SoundEvents.EVOKER_CAST_SPELL, SoundSource.PLAYERS, 0.4F, level.getRandom().nextFloat() * 1.2F + 0.6F);

    var text = compound.getList("pages", Tag.TAG_STRING);
    Spell spell = DictionarySavedData.computeIfAbsent(server).parse(LimitedBookSerializer.decodeText(text));
    if(spell != null) {
      if(spell.cost() > level.getGameRules().getInt(ScriptorGameRules.TOME_MAX_COST)) {
        player.sendSystemMessage(Component.translatable("extra.scriptor.fizzle"));
        ScriptorAdvancements.FIZZLE.get().trigger((ServerPlayer) player);
        if(!player.isCreative())
          addCooldown(player, 350);
        return true;
      }
      spell.cast(new SpellbookTargetable(itemStack, player, player.getInventory().selected).withTargetItem(false));
      if(!player.isCreative())
        addCooldown(player, (int) Math.round(spell.cost() * 7));
      return false;
    }
    return true;
  }

  static void addCooldown(Player player, int time) {
    for(var spellbook: SPELLBOOKS)
      player.getCooldowns().addCooldown(spellbook, time);
  }
}

File: common/src/main/kotlin/com/ssblur/scriptor/helpers/SpellbookHelper.kt
package com.ssblur.scriptor.helpers

import com.ssblur.scriptor.advancement.ScriptorAdvancements
import com.ssblur.scriptor.config.ScriptorConfig
import com.ssblur.scriptor.data.components.ConditionData
import com.ssblur.scriptor.data.components.ReagentData
import com.ssblur.scriptor.data.components.ScriptorDataComponents
import com.ssblur.scriptor.data.saved_data.DictionarySavedData.Companion.computeIfAbsent
import com.ssblur.scriptor.effect.EmpoweredStatusEffect
import com.ssblur.scriptor.extension.EntityCastCooldownExtension.canCast
import com.ssblur.scriptor.extension.EntityCastCooldownExtension.castCooldown
import com.ssblur.scriptor.helpers.LimitedBookSerializer.decodeText
import com.ssblur.scriptor.helpers.targetable.SpellbookTargetable
import com.ssblur.scriptor.helpers.targetable.Targetable
import com.ssblur.scriptor.word.ConditionalWord
import com.ssblur.scriptor.word.descriptor.discount.ReagentDescriptor
import com.ssblur.scriptor.word.subject.HitSubject
import net.minecraft.core.component.DataComponents
import net.minecraft.core.registries.BuiltInRegistries
import net.minecraft.network.chat.Component
import net.minecraft.resources.ResourceLocation
import net.minecraft.server.level.ServerLevel
import net.minecraft.server.level.ServerPlayer
import net.minecraft.world.entity.Entity
import net.minecraft.world.entity.player.Player
import net.minecraft.world.item.ItemStack
import kotlin.math.roundToInt

object SpellbookHelper {
  fun isInventoryCaster(
    itemStack: ItemStack,
    level: ServerLevel
  ): Boolean {
    val text = itemStack.get(DataComponents.WRITTEN_BOOK_CONTENT) ?: return false
    val spell = computeIfAbsent(level).parse(decodeText(text))
    return (spell != null && (spell.subject?.canBeCastOnInventory() == true))
  }

  fun getReagentData(
    itemStack: ItemStack,
    level: ServerLevel
  ): ReagentData {
    val text = itemStack.get(DataComponents.WRITTEN_BOOK_CONTENT) ?: return ReagentData.NONE
    val spell = computeIfAbsent(level).parse(decodeText(text)) ?: return ReagentData.NONE
    val map = mutableMapOf<ResourceLocation, Int>()
    spell.spells.forEach {
      it.descriptors.forEach { descriptor ->
        if(descriptor is ReagentDescriptor) {
          val key = BuiltInRegistries.ITEM.getKey(descriptor.item)
          map[key] = map[key] ?: 0
          map[key] = map[key]!! + 1
        }
      }
    }
    return ReagentData(map)
  }

  fun getConditionData(
    itemStack: ItemStack,
    level: ServerLevel
  ): ConditionData {
    val text = itemStack.get(DataComponents.WRITTEN_BOOK_CONTENT) ?: return ConditionData.NONE
    val data = computeIfAbsent(level)
    val spell = data.parse(decodeText(text)) ?: return ConditionData.NONE
    val list = mutableListOf<String>()

    spell.subject.let { subject ->
      if(subject is ConditionalWord) {
        data.getKey(subject)?.replace(":", ".")?.let { condition ->
          list.add(condition)
        }
      }
    }

    spell.spells.forEach {
      it.descriptors.forEach { descriptor ->
        if(descriptor is ConditionalWord) {
          data.getKey(descriptor)?.replace(":", ".")?.let { condition ->
            list.add(condition)
          }
        }
      }

      it.action.let { action ->
        if(action is ConditionalWord) {
          data.getKey(action)?.replace(":", ".")?.let { condition ->
            list.add(condition)
          }
        }
      }
    }
    return ConditionData(list)
  }

  fun castFromItem(
    itemStack: ItemStack,
    player: Player,
    maxCost: Int? = null,
    costMultiplier: Int? = null,
    cooldownFunc: (Player, Int) -> Unit = ::addCooldown,
    targetOverride: List<Targetable>? = null,
  ): Boolean {
    val adjustedCostMultiplier = (costMultiplier ?: ScriptorConfig.TOME_COOLDOWN_MULTIPLIER()).toDouble() / 100.0
    val level = player.level()
    if (level !is ServerLevel) return false
    val spell = computeIfAbsent(level).parse(
      itemStack[DataComponents.WRITTEN_BOOK_CONTENT]?.let {
        decodeText(it)
      } ?: itemStack[ScriptorDataComponents.SPELL]
    )

    if (spell != null && spell.subject !is HitSubject) {
      spell.deduplicatedDescriptorsForSubjects()
      spell.playSound(level, player.blockPosition())
      if (!player.canCast(spell, adjustedCostMultiplier * 49)) {
        player.sendSystemMessage(Component.translatable("extra.scriptor.fizzle"))
        ScriptorAdvancements.FIZZLE.get().trigger(player as ServerPlayer)
        if (!player.isCreative)
          cooldownFunc(player, (350 * adjustedCostMultiplier).roundToInt())
        return true
      }

      val subject = SpellbookTargetable(itemStack, player, player.inventory.selected)
        .withTargetItem(false)
      if(targetOverride == null)
        spell.cast(subject)
      else
        spell.castOnTargets(subject, targetOverride, castHooks = true)

      if (!player.isCreative) {
        var costScale = 1.0
        for (instance in player.activeEffects)
          if (instance.effect.value() is EmpoweredStatusEffect)
            (0..instance.amplifier).forEach{ _ ->
              costScale *= (instance.effect.value() as EmpoweredStatusEffect).scale.toDouble()
            }
        val adjustedCost =
          costScale * spell.cost() * adjustedCostMultiplier
        cooldownFunc(player, (adjustedCost * 7).roundToInt())
        return true
      }
      return false
    }
    return false
  }

  fun addCooldown(entity: Entity, time: Int) {
    if(entity !is Player || !entity.isCreative) entity.castCooldown = time.toLong() * 7L
  }
}
